File: test/arguments/with_range_check_code/cpp98/VarUInt32RangeCheckTest.cpp
#include "gtest/gtest.h"

#include "zserio/CppRuntimeException.h"

#include "with_range_check_code/varuint32_range_check/VarUInt32RangeCheckCompound.h"

namespace with_range_check_code
{
namespace varuint32_range_check
{

class VarUInt32RangeCheckTest : public ::testing::Test
{
protected:
    void checkVarUInt32Value(uint32_t value)
    {
        VarUInt32RangeCheckCompound varUInt32RangeCheckCompound;
        varUInt32RangeCheckCompound.setValue(value);
        zserio::BitStreamWriter writer;
        varUInt32RangeCheckCompound.write(writer);
        size_t writeBufferByteSize;
        const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
        zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);
        const VarUInt32RangeCheckCompound readVarUInt32RangeCheckCompound(reader);
        ASSERT_EQ(varUInt32RangeCheckCompound, readVarUInt32RangeCheckCompound);
    }

    static const uint32_t   VARUINT32_LOWER_BOUND = UINT32_C(0);
    static const uint32_t   VARUINT32_UPPER_BOUND = (UINT32_C(1) << 29) - 1;
};

TEST_F(VarUInt32RangeCheckTest, varUInt32LowerBound)
{
    checkVarUInt32Value(VARUINT32_LOWER_BOUND);
}

TEST_F(VarUInt32RangeCheckTest, varUInt32UpperBound)
{
    checkVarUInt32Value(VARUINT32_UPPER_BOUND);
}

TEST_F(VarUInt32RangeCheckTest, varUInt32AboveUpperBound)
{
    try
    {
        checkVarUInt32Value(VARUINT32_UPPER_BOUND + 1);
        FAIL() << "Actual: no exception, Expected: zserio::CppRuntimeException";
    }
    catch (const zserio::CppRuntimeException& excpt)
    {
        ASSERT_STREQ("Value 536870912 of VarUInt32RangeCheckCompound.value exceeds the range of "
                "<0..536870911>!", excpt.what());
    }
}

} // namespace varuint32_range_check
} // namespace with_range_check_code

File: test/arguments/with_range_check_code/java/with_range_check_code/VarUInt32RangeCheckTest.java
package with_range_check_code;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import with_range_check_code.varuint32_range_check.VarUInt32RangeCheckCompound;

import java.io.IOException;

import zserio.runtime.ZserioError;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

public class VarUInt32RangeCheckTest
{
    @Test
    public void varUInt32LowerBound() throws IOException, ZserioError
    {
        checkVarUInt32Value(VARUINT32_LOWER_BOUND);
    }

    @Test
    public void varUInt32UpperBound() throws IOException, ZserioError
    {
        checkVarUInt32Value(VARUINT32_UPPER_BOUND);
    }

    @Test
    public void varUInt32BelowLowerBound() throws IOException, ZserioError
    {
        assertThrows(ZserioError.class, () -> checkVarUInt32Value(VARUINT32_LOWER_BOUND - 1));
    }

    @Test
    public void varUInt32AboveUpperBound() throws IOException, ZserioError
    {
        assertThrows(ZserioError.class, () -> checkVarUInt32Value(VARUINT32_UPPER_BOUND + 1));
    }

    private void checkVarUInt32Value(int value) throws IOException, ZserioError
    {
        VarUInt32RangeCheckCompound varUInt32RangeCheckCompound = new VarUInt32RangeCheckCompound();
        varUInt32RangeCheckCompound.setValue(value);
        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        varUInt32RangeCheckCompound.write(writer);
        final byte[] writtenByteArray = writer.toByteArray();
        writer.close();

        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(writtenByteArray);
        final VarUInt32RangeCheckCompound readVarUInt32RangeCheckCompound =
                new VarUInt32RangeCheckCompound(reader);
        assertEquals(varUInt32RangeCheckCompound, readVarUInt32RangeCheckCompound);
    }

    private static final int VARUINT32_LOWER_BOUND = 0;
    private static final int VARUINT32_UPPER_BOUND = 536870911;
}
